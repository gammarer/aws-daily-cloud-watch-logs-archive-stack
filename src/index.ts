import * as crypto from 'crypto';
import { SecureLogBucket } from '@yicr/secure-log-bucket';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as scheduler from 'aws-cdk-lib/aws-scheduler';
import { Construct } from 'constructs';
import { LogArchiverFunction } from './funcs/log-archiver-function';

export interface DailyCloudWatchLogArchiverProps {
  readonly schedules: ScheduleProperty[];
}

export interface ScheduleProperty {
  readonly name: string;
  readonly description: string;
  readonly target: ScheduleTargetProperty;
}

export interface ScheduleTargetProperty {
  readonly logGroupName: string;
  readonly destinationPrefix: string;
}

export class DailyCloudWatchLogArchiver extends Construct {
  constructor(scope: Construct, id: string, props: DailyCloudWatchLogArchiverProps) {
    super(scope, id);

    // props validation
    if (props.schedules.length === 0) {
      throw new Error('Schedule not set.');
    }
    if (props.schedules.length > 60) {
      throw new Error('Maximum number(60) of schedule.');
    }

    // ðŸ‘‡Get current account & region
    //const account = cdk.Stack.of(this).account;
    //const stackName: string = cdk.Stack.of(this).stackName;
    const region = cdk.Stack.of(this).region;

    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`)
      .digest('hex');

    // ðŸ‘‡Create Backup S3 Bucket
    const logArchiveBucket = new SecureLogBucket(this, 'LogArchiveBucket', {
      bucketName: `log-archive-${randomNameKey}`,
    });
    logArchiveBucket.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.ServicePrincipal(`logs.${region}.amazonaws.com`),
      ],
      actions: [
        's3:GetBucketAcl',
      ],
      resources: [
        logArchiveBucket.bucketArn,
      ],
    }));
    logArchiveBucket.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        new iam.ServicePrincipal(`logs.${region}.amazonaws.com`),
      ],
      actions: [
        's3:PutObject',
      ],
      resources: [
        `${logArchiveBucket.bucketArn}/*`,
      ],
      conditions: {
        StringEquals: {
          's3:x-amz-acl': 'bucket-owner-full-control',
        },
      },
    }));

    // ðŸ‘‡Create Lambda Execution role.
    const lambdaExecutionRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `daily-cw-log-archiver-lambda-exec-${randomNameKey}-role`,
      description: '',
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        ['log-export-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'logs:CreateExportTask',
              ],
              resources: ['*'],
            }),
          ],
        }),
        ['put-bucket-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetBucketAcl',
                's3:PutObject',
              ],
              resources: [
                logArchiveBucket.bucketArn,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Create Lambda Function
    const lambdaFunction = new LogArchiverFunction(this, 'LogArchiveFunction', {
      functionName: `daily-cw-log-archiver-${randomNameKey}-func`,
      description: 'A function to archive logs s3 bucket from CloudWatch Logs.',
      environment: {
        BUCKET_NAME: logArchiveBucket.bucketName,
      },
      role: lambdaExecutionRole,
    });

    // ðŸ‘‡EventBridge Scheduler IAM Role
    const schedulerExecutionRole = new iam.Role(this, 'SchedulerExecutionRole', {
      assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
      inlinePolicies: {
        ['lambda-invoke-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: [
                lambdaFunction.functionArn,
                `${lambdaFunction.functionArn}:*`,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Create Schedule Group
    const scheduleGroup = new scheduler.CfnScheduleGroup(this, 'ScheduleGroup', {
      name: `log-archive-schedule-${randomNameKey}-group`,
    });

    for (const [index, schedule] of Object.entries(props.schedules)) {
      // ðŸ‘‡Schedule ID prefix
      const idPrefix = crypto.createHash('shake256', { outputLength: 4 })
        .update(schedule.name)
        .digest('hex');

      // ðŸ‘‡Schedule
      new scheduler.CfnSchedule(this, `Schedule${idPrefix}`, {
        name: schedule.name,
        description: schedule.description,
        state: 'ENABLED',
        groupName: scheduleGroup.name,
        flexibleTimeWindow: {
          mode: 'OFF',
        },
        scheduleExpressionTimezone: 'UTC',
        scheduleExpression: `cron(${index} 13 * * ? *)`, // max 60
        target: {
          arn: lambdaFunction.functionArn,
          roleArn: schedulerExecutionRole.roleArn,
          input: JSON.stringify({
            logGroupName: schedule.target.logGroupName,
            destinationPrefix: schedule.target.destinationPrefix,
          }),
          retryPolicy: {
            maximumEventAgeInSeconds: 60,
            maximumRetryAttempts: 0,
          },
        },
      });
    }
  }
}