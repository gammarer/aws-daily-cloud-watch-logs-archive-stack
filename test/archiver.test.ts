import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { DailyCloudWatchLogsArchiver } from '../src';

describe('DailyCloudWatchLogsArchiver Testing', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  new DailyCloudWatchLogsArchiver(stack, 'DailyCloudWatchLogsArchiver', {
    schedules: [
      {
        name: 'example-log-archive-1st-rule',
        description: 'example log archive 1st rule.',
        target: {
          logGroupName: 'example-log-1st-group',
          destinationPrefix: 'example-1st-log',
        },
      },
      {
        name: 'example-log-archive-2nd-rule',
        description: 'example log archive 2nd rule.',
        target: {
          logGroupName: 'example-log-2nd-group',
          destinationPrefix: 'example-2nd-log',
        },
      },
    ],
  });

  const template = Template.fromStack(stack);

  describe('Bucket Testing', () => {

    it('Should have bucket encryption', () => {
      template.hasResourceProperties('AWS::S3::Bucket', {
        BucketEncryption: Match.objectEquals({
          ServerSideEncryptionConfiguration: [
            {
              ServerSideEncryptionByDefault: {
                SSEAlgorithm: 'aws:kms',
              },
            },
          ],
        }),
      });
    });

    it('Should have bucket resource policy', () => {
      template.hasResourceProperties('AWS::S3::BucketPolicy', {
        PolicyDocument: {
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:GetBucketAcl',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLogArchiveBucket'),
                  'Arn',
                ],
              },
            }),
            Match.objectEquals({
              Effect: 'Allow',
              Action: 's3:PutObject',
              Principal: {
                Service: 'logs.us-east-1.amazonaws.com',
              },
              Resource: {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLogArchiveBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
              Condition: {
                StringEquals: {
                  's3:x-amz-acl': 'bucket-owner-full-control',
                },
              },
            }),
          ]),
        },
      });
    });

  });

  describe('Lambda Testing', () => {

    it('Should have lambda execution role', () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('daily-cw-log-archiver-lambda-exec-.*-role'),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'lambda.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        ManagedPolicyArns: Match.arrayWith([
          {
            'Fn::Join': Match.arrayEquals([
              '',
              Match.arrayEquals([
                'arn:',
                {
                  Ref: 'AWS::Partition',
                },
                ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
              ]),
            ]),
          },
        ]),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'log-export-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: 'logs:CreateExportTask',
                  Resource: '*',
                }),
              ],
            }),
          },
          {
            PolicyName: 'put-bucket-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayEquals([
                    's3:GetBucketAcl',
                    's3:PutObject',
                  ]),
                  Resource: {
                    'Fn::GetAtt': Match.arrayEquals([
                      Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLogArchiveBucket.*'),
                      'Arn',
                    ]),
                  },
                }),
              ],
            }),
          },
        ]),
      }));
    });

    it('Should have lambda function', () => {
      template.hasResourceProperties('AWS::Lambda::Function', Match.objectEquals({
        FunctionName: Match.stringLikeRegexp('daily-cw-log-archiver-.*-func'),
        Handler: 'index.handler',
        Runtime: 'nodejs18.x',
        Code: {
          S3Bucket: Match.anyValue(),
          S3Key: Match.stringLikeRegexp('.*.zip'),
        },
        Description: Match.anyValue(),
        Environment: {
          Variables: {
            AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
            BUCKET_NAME: {
              Ref: Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLogArchiveBucket.*'),
            },
          },
        },
        Role: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLambdaExecutionRole.*'),
            'Arn',
          ],
        },
      }));
    });
  });

  // todo: scheduler property.
  it('Should have Schedule', () => {
    template.hasResourceProperties('AWS::Scheduler::ScheduleGroup', Match.objectEquals({
      Name: Match.stringLikeRegexp('log-archive-schedule-.*-group'),
    }));
    template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
      Name: Match.anyValue(),
      Description: Match.anyValue(),
      GroupName: Match.stringLikeRegexp('log-archive-schedule-.*-group'),
      State: 'ENABLED',
      FlexibleTimeWindow: {
        Mode: 'OFF',
      },
      ScheduleExpressionTimezone: 'UTC',
      ScheduleExpression: Match.stringLikeRegexp('cron(.* 13 * * ? *)'),
      Target: Match.objectEquals({
        Arn: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogsArchiverLogArchiveFunction.*'),
            'Arn',
          ],
        },
        RoleArn: {
          'Fn::GetAtt': [
            Match.stringLikeRegexp('DailyCloudWatchLogsArchiverSchedulerExecutionRole.*'),
            'Arn',
          ],
        },
        Input: Match.stringLikeRegexp('{"logGroupName":"example-log-.*-group","destinationPrefix":"example-.*-log"}'),
        RetryPolicy: {
          MaximumEventAgeInSeconds: 60,
          MaximumRetryAttempts: 0,
        },
      }),
    }));
    template.resourceCountIs('AWS::Scheduler::Schedule', 2);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot('archiver');
  });

  describe('DailyCloudWatchLogArchiver Error Handling Testing', () => {
    it('Should have error of schedule not set', () => {
      expect(() => {
        new DailyCloudWatchLogsArchiver(new Stack(new App()), 'DailyCloudWatchLogsArchiver', {
          schedules: [],
        });
      }).toThrow(Error);
    });
    it('Should have error of schedule count over', () => {
      expect(() => {
        new DailyCloudWatchLogsArchiver(new Stack(new App()), 'DailyCloudWatchLogsArchiver', {
          schedules: [...Array(61)].map((_, i) => {
            const id = ('00' + i).slice(-2);
            return {
              name: `example-${id}-schedule`,
              description: `example ${id} schedule`,
              target: {
                logGroupName: `example-log-${id}-group`,
                destinationPrefix: `example-${id}-log`,
              },
            };
          }),
        });
      }).toThrow(Error);
    });
  });
});
