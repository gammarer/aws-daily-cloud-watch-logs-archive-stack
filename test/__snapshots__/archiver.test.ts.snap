// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyCloudWatchLogsArchiver Testing Should match snapshot: archiver 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DailyCloudWatchLogsArchiverLambdaExecutionRoleB287F490": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "logs:CreateExportTask",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "log-export-policy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "s3:GetBucketAcl",
                    "s3:PutObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "put-bucket-policy",
          },
        ],
        "RoleName": "daily-cw-log-archiver-lambda-exec-a5ac343a-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "DailyCloudWatchLogsArchiverLogArchiveBucket88461686": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "log-archive-a5ac343a",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Id": "archive-step-lifecycle-rule",
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 60,
                },
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 120,
                },
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 180,
                },
                Object {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 360,
                },
              ],
            },
          ],
        },
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerEnforced",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "DailyCloudWatchLogsArchiverLogArchiveBucketPolicy098686E8": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "StringEquals": Object {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "logs.us-east-1.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "DailyCloudWatchLogsArchiverLogArchiveFunction1589C8E4": Object {
      "DependsOn": Array [
        "DailyCloudWatchLogsArchiverLambdaExecutionRoleB287F490",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "A function to archive logs s3 bucket from CloudWatch Logs.",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Ref": "DailyCloudWatchLogsArchiverLogArchiveBucket88461686",
            },
          },
        },
        "FunctionName": "daily-cw-log-archiver-a5ac343a-func",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DailyCloudWatchLogsArchiverLambdaExecutionRoleB287F490",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DailyCloudWatchLogsArchiverSchedule0c256a4b0CD0D96F": Object {
      "Properties": Object {
        "Description": "example log archive 2nd rule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "GroupName": "log-archive-schedule-a5ac343a-group",
        "Name": "example-log-archive-2nd-rule",
        "ScheduleExpression": "cron(1 13 * * ? *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": Object {
          "Arn": Object {
            "Fn::GetAtt": Array [
              "DailyCloudWatchLogsArchiverLogArchiveFunction1589C8E4",
              "Arn",
            ],
          },
          "Input": "{"logGroupName":"example-log-2nd-group","destinationPrefix":"example-2nd-log"}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "DailyCloudWatchLogsArchiverSchedulerExecutionRoleB476F543",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "DailyCloudWatchLogsArchiverSchedule78b9c1a9F3486471": Object {
      "Properties": Object {
        "Description": "example log archive 1st rule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "GroupName": "log-archive-schedule-a5ac343a-group",
        "Name": "example-log-archive-1st-rule",
        "ScheduleExpression": "cron(0 13 * * ? *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": Object {
          "Arn": Object {
            "Fn::GetAtt": Array [
              "DailyCloudWatchLogsArchiverLogArchiveFunction1589C8E4",
              "Arn",
            ],
          },
          "Input": "{"logGroupName":"example-log-1st-group","destinationPrefix":"example-1st-log"}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "DailyCloudWatchLogsArchiverSchedulerExecutionRoleB476F543",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "DailyCloudWatchLogsArchiverScheduleGroupF93344C1": Object {
      "Properties": Object {
        "Name": "log-archive-schedule-a5ac343a-group",
      },
      "Type": "AWS::Scheduler::ScheduleGroup",
    },
    "DailyCloudWatchLogsArchiverSchedulerExecutionRoleB476F543": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "lambda:InvokeFunction",
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "DailyCloudWatchLogsArchiverLogArchiveFunction1589C8E4",
                        "Arn",
                      ],
                    },
                    Object {
                      "Fn::Join": Array [
                        "",
                        Array [
                          Object {
                            "Fn::GetAtt": Array [
                              "DailyCloudWatchLogsArchiverLogArchiveFunction1589C8E4",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "lambda-invoke-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
